File: swift/Tasks/Data/DittoManager.swift
import Foundation
import DittoSwift

// MARK: DataManager Protocol
@MainActor protocol DataManager : ObservableObject {
    var appManager: AppManager { get set}
    var tasks: [TaskModel] { get set }
    var isPresentingEditScreen: Bool { get set }
    var selectedTaskModel: TaskModel? { get set }
    
    func initialize() async throws
    func populateTaskCollection() async throws
    func setSyncEnabled(_ newValue: Bool) throws
    func insertTaskModel(_ task: TaskModel) async
    func updateTaskModel(_ task: TaskModel) async
    func toggleComplete(task: TaskModel) async
    func deleteTaskModel(_ task: TaskModel) async
    
    
}

// MARK: DittoManager Implementation
@MainActor class DittoManager : ObservableObject, DataManager {
    @Published var tasks = [TaskModel]()
    @Published var isPresentingEditScreen: Bool = false
    @Published var selectedTaskModel: TaskModel?
    
    var subscription: DittoSyncSubscription?
    var storeObserver: DittoStoreObserver?
    var ditto: Ditto?
    
    var appManager: AppManager
    
    init(appManager: AppManager) {
        //cache state for future use
        self.appManager = appManager
    }
    
    func initialize() async throws {
        // setup logging
        DittoLogger.enabled = true
        
        //setup logging level
        let isPreview: Bool =
        ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1"
        if !isPreview {
            DittoLogger.minimumLogLevel = .debug
        }
        
        //
        //TODO: setup Ditto Identity
        //
        //UPDATE CODE HERE
        
        do {
            //Disable sync with v3 peers, required for DQL
            try ditto?.disableSyncWithV3()
            
            //setup the collection with initial data and then setup observer
            try await self.populateTaskCollection()
            try self.registerObservers()
            
        } catch let error {
            self.appManager.setError(error)
        }
    }
    
    /// Performs cleanup of Ditto resources
    ///
    /// This method handles the graceful shutdown of Ditto components by:
    /// - Cancelling any active subscriptions
    /// - Cancelling store observers
    /// - Stopping the Ditto sync process
    deinit {
        subscription?.cancel()
        subscription = nil
        
        storeObserver?.cancel()
        storeObserver = nil
        
        if let dittoInstance = ditto {
            if (dittoInstance.isSyncActive) {
                dittoInstance.stopSync()
            }
        }
    }
    
}

// MARK:  Populate Initial Tasks for Task Collection
extension DittoManager {
    
    /// Populates the Ditto tasks collection with initial seed data if it's empty
    ///
    /// This method creates a set of predefined tasks in the Ditto store by:
    /// - Defining an array of initial `TaskModel` objects with unique IDs and titles
    /// - Inserting each task into the Ditto store using DQL (Ditto Query Language)
    /// - Using the INITIAL keyword to only insert if documents don't already exist
    ///
    /// The initial tasks include common todo items like:
    /// - Buy groceries
    /// - Clean the kitchen
    /// - Schedule dentist appointment
    /// - Pay bills
    ///
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/read#using-args-to-query-dynamic-values///
    ///
    /// - Throws: A DittoError if the insert operations fail
    func populateTaskCollection() async throws {
        
        let initialTasks: [TaskModel] = [
            TaskModel(
                _id: "50191411-4C46-4940-8B72-5F8017A04FA7",
                title: "Buy groceries"),
            TaskModel(
                _id: "6DA283DA-8CFE-4526-A6FA-D385089364E5",
                title: "Clean the kitchen"),
            TaskModel(
                _id: "5303DDF8-0E72-4FEB-9E82-4B007E5797F0",
                title: "Schedule dentist appointment"),
            TaskModel(
                _id: "38411F1B-6B49-4346-90C3-0B16CE97E174",
                title: "Pay bills"),
        ]
        
        for task in initialTasks {
            do {
                if let dittoInstance = ditto {
                    //
                    //TODO: add tasks into the ditto collection using INSERT statment
                    // https://docs.ditto.live/sdk/latest/crud/write#inserting-documents
                    //
                    
                    //UPDATE CODE HERE
                }
            } catch {
                appManager.setError(error)
            }
        }
    }
}


// MARK: Register Observer - Live Query
extension DittoManager {
    /// Registers observers for the planets collection to handle real-time updates.
    ///
    /// This method sets up a live query observer that:
    /// - Monitors the tasks collection for changes
    /// - Updates the @Published tasks array when changes occur
    /// - Filters out tasks NOT deleted
    ///
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/read
    ///
    /// - Throws: A DittoError if the observer cannot be registered
    func registerObservers() throws {
        if let dittoInstance = ditto {
            
            //
            //TODO: - setup observer query, filter out NOT deleted
            //
            
            //UPDATE CODE HERE
            
            //
            //TODO: - setup store observer with query and set array with TaskModel
            //
            
            //UPDATE CODE HERE
        }
    }
}

// MARK: Subscriptions and Sync

/// Sets up the initial subscription to the tasks collection in Ditto.
///
/// This subscription ensures that changes to the tasks collection are synced
/// between the local Ditto store the BigPeer or other SmallPeers
///
/// - SeeAlso: https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
///
extension DittoManager {
    
    func setSyncEnabled(_ newValue: Bool) throws {
        if let dittoInstance = ditto {
            if dittoInstance.isSyncActive && newValue {
                try startSync()
            } else if dittoInstance.isSyncActive && !newValue {
                stopSync()
            }
        }
    }
    
    private func startSync() throws {
        do {
            if let dittoInstance = ditto {
                //
                //TODO: implement the startSync
                // https://docs.ditto.live/sdk/latest/install-guides/swift#integrating-and-initializing-sync
                //
                
                //UPDATE CODE HERE
                
                //
                //TODO: implement the set subscription
                // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
                //
                
                //UPDATE CODE HERE
            }
        } catch {
            appManager.setError(error)
            throw error
        }
    }
    
    private func stopSync () {
        //
        //TODO: implement stopping sync by calling cancel and setting object to nil
        //and then running stopSync function
        // https://docs.ditto.live/sdk/latest/sync/syncing-data#canceling-subscriptions
        //
        subscription?.cancel()
        subscription = nil
        
        ditto?.stopSync()
    }
}

// MARK: Adding/Editing Tasks
extension DittoManager {
    
    /// Creates a new TaskModel document in the Ditto store.
    ///
    /// This method:
    /// - Creates a new document in the tasks collection
    ///
    /// - Parameter task: The TaskModel to add to the store
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/create#creating-documents
    ///
    /// - Throws: A DittoError if the insert operation fails
    func insertTaskModel(_ task: TaskModel) async {
        let newTask = task.value
        //
        // TODO - write INSERT DQL Statement
        // https://docs.ditto.live/dql/insert
        // https://docs.ditto.live/sdk/latest/crud/create#creating-documents
        //
        
        //UPDATE CODE HERE
        let query = ""
        
        do {
            if let dittoInstance = ditto {
                //
                //TODO: use dittoInstance store to execute DQL with arguments
                // https://docs.ditto.live/sdk/latest/crud/create#creating-documents
                //
                
                //UPDATE CODE HERE
            }
        } catch {
            appManager.setError(error)
        }
    }
    
    /// Updates an existing TaskModel's properties in the Ditto store.
    ///
    /// This method uses DQL to update all mutable fields of the TaskModel
    ///
    /// - Parameter planet: The TaskModel object containing the updated values
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/update#updating
    ///
    /// - Throws: A DittoError if the update operation fails
    func updateTaskModel(_ task: TaskModel) async {
        //
        //TODO: write UPDATE DQL Statement
        // https://docs.ditto.live/dql/update#basic-update
        // https://docs.ditto.live/sdk/latest/crud/update#updating
        //
        
        //UPDATE CODE HERE
        let query = ""
        
        do {
            if let dittoInstance = ditto {
                //
                //TODO: use dittoInstance store to execute DQL with arguments
                // https://docs.ditto.live/sdk/latest/crud/update#updating
                //
                
                //UPDATE CODE HERE
            }
        } catch {
            appManager.setError(error)
        }
        
    }
    
    
}

// MARK: Updating TaskModel Complete
extension DittoManager {
    
    /// Toggles the completion status of a task in the Ditto store
    ///
    /// This method:
    /// - Inverts the current 'done' status of the task
    /// - Updates only the 'done' field in the store using DQL
    /// - Maintains all other task properties unchanged
    ///
    /// - Parameter task: The TaskModel to toggle completion status for
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/update#updating
    ///
    /// - Throws: A DittoError if the update operation fails
    func toggleComplete(task: TaskModel) async {
        let done = !task.done
        
        //
        //TODO: write UPDATE DQL Statement
        // https://docs.ditto.live/dql/update#basic-update
        // https://docs.ditto.live/sdk/latest/crud/update#updating
        //
        
        //UPDATE CODE HERE
        let query = ""
        
        do {
            if let dittoInstance = ditto {
                //
                //TODO: use dittoInstance store to execute DQL with arguments
                // https://docs.ditto.live/sdk/latest/crud/update#updating
                //
                
                // UPDATE CODE HERE
            }
        } catch {
            appManager.setError(error)
        }
    }
}

// MARK: Deleting TaskModel
extension DittoManager {
    
    /// Delete a TaskModel by setting its deleted flag to true.
    ///
    /// This method implements the 'Soft-Delete' pattern, which:
    /// - Marks the TaskModel as deleted instead of evicting it
    /// - Removes it from active queries and views
    /// - Maintains the data for historical purposes
    ///
    /// - Parameter task: The TaskModel to soft-delet
    /// - SeeAlso: https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
    ///
    /// - Throws: A DittoError if the archive operation fails
    func deleteTaskModel(_ task: TaskModel) async {
        //
        //TODO: write UPDATE DQL Statement using Soft-Delete pattern
        // https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
        //
        
        //UPDATE CODE HERE
        let query = ""
        do {
            if let dittoInstance = ditto {
                //
                //TODO: use dittoInstance store to execute DQL with arguments
                // https://docs.ditto.live/sdk/latest/crud/update#updating
                //
                
                //UPDATE CODE HERE
            }
        } catch {
            appManager.setError(error)
        }
    }
}

File: course-101/android/app/src/main/java/live/ditto/quickstart/tasks/data/DittoManager.kt
package live.ditto.quickstart.tasks.data

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.withContext
import live.ditto.Ditto
import live.ditto.DittoError
import live.ditto.DittoIdentity
import live.ditto.DittoLogLevel
import live.ditto.DittoLogger
import live.ditto.DittoStoreObserver
import live.ditto.DittoSyncSubscription
import live.ditto.android.AndroidDittoDependencies
import live.ditto.quickstart.tasks.models.DittoConfig
import live.ditto.quickstart.tasks.models.TaskModel
import live.ditto.quickstart.tasks.services.ErrorService

class DittoManager(
    override val dittoConfig: DittoConfig,
    androidDittoDependencies: AndroidDittoDependencies,
    private val errorService: ErrorService
) : DataManager {
    companion object {
        private const val TAG = "DittoManager"
    }

    var ditto: Ditto? = null
    private var subscription: DittoSyncSubscription? = null
    private var storeObserver: DittoStoreObserver? = null

    init {
        try {
            DittoLogger.minimumLogLevel = DittoLogLevel.DEBUG

            //
            //TODO: setup Ditto Identity
            //
            //UPDATE CODE HERE

            // disable sync with v3 peers, required for syncing with the Ditto Cloud (Big Peer)
            this.ditto?.disableSyncWithV3()

        } catch (e: DittoError) {
            errorService.showError("Failed to initialize Ditto: ${e.message}")
        }
    }

    /**
     * Populates the Ditto tasks collection with initial seed data if it's empty.
     * 
     * This method:
     * - Creates a set of predefined tasks with unique IDs and titles
     * - Inserts each task into the Ditto store using DQL (Ditto Query Language)
     * - Uses the INITIAL keyword to only insert if documents don't already exist
     * 
     * The initial tasks include common todo items like:
     * - Buy groceries
     * - Clean the kitchen
     * - Schedule dentist appointment
     * - Pay bills
     * 
     * Implementation details:
     * - Uses DQL INSERT statement with INITIAL keyword to prevent duplicates
     * - Each task document contains _id, title, done, and deleted fields
     * - Handles errors gracefully and logs them for debugging
     * 
     * @throws Exception if the insert operations fail
     * Ditto Docs:
     * https://docs.ditto.live/sdk/latest/crud/write#inserting-documents
     */
    override suspend fun populateTaskCollection() {
        return withContext(Dispatchers.IO) {
            val tasks = listOf(
                TaskModel("50191411-4C46-4940-8B72-5F8017A04FA7", "Buy groceries"),
                TaskModel("6DA283DA-8CFE-4526-A6FA-D385089364E5", "Clean the kitchen"),
                TaskModel("5303DDF8-0E72-4FEB-9E82-4B007E5797F0", "Schedule dentist appointment"),
                TaskModel("38411F1B-6B49-4346-90C3-0B16CE97E174", "Pay bills")
            )

            tasks.forEach { task ->
                try {
                    //
                    //TODO: add tasks into the ditto collection using INSERT statment
                    // https://docs.ditto.live/sdk/latest/crud/write#inserting-documents
                    //

                    //UPDATE CODE HERE
                } catch (e: Exception) {
                    errorService.showError("Unable to insert initial documents: ${e.message}")
                    Log.e(TAG, "Unable to insert initial document", e)
                }
            }
        }
    }

    /**
     * Creates a Flow that observes and emits changes to the tasks collection in Ditto.
     * https://docs.ditto.live/sdk/latest/crud/read#using-args-to-query-dynamic-values
     *
     * This method:
     * - Sets up a live query observer for the tasks collection
     * - Emits a new list of TaskModel objects whenever the data changes
     * - Automatically cleans up the observer when the flow is cancelled
     *
     * The query:
     * - Returns all non-archived TaskModel (done = false)
     * - Transforms raw Ditto data into TaskModel objects
     *
     * @return Flow<List<TaskModel>> A flow that emits updated lists of Tasks
     * whenever changes occur in the collection
     *
     */
    override fun getTaskModels(): Flow<List<TaskModel>> = callbackFlow {
        try {
            //
            //TODO: - setup observer query, filter out NOT deleted
            //

            //UPDATE CODE HERE
            val query = ""

            //
            //TODO: - setup store observer with query and set array with TaskModel
            //

            //UPDATE CODE HERE

            awaitClose {
                //
                //TODO: - Clean up the observer when the flow is cancelled
                //

                //UPDATE CODE HERE
                storeObserver?.close()
                storeObserver = null
            }
        } catch (e: Exception) {
            errorService.showError("Failed to setup observer for getting taskModels: ${e.message}")
            Log.e(TAG, "Failed to setup observer for getting taskModels", e)
        }
    }

    /**
     * Enables or disables synchronization with the Ditto cloud (BigPeer).
     * 
     * This method:
     * - Manages the sync state of the Ditto instance
     * - Persists the sync preference in DataStore
     * - Updates the UI state through LiveData
     * 
     * Implementation details:
     * - Retrieves default sync state from preferences if not specified
     * - Starts sync if enabled and not already active
     * - Stops sync if disabled and currently active
     * - Updates both DataStore and LiveData to maintain state
     * 
     * @param enabled Boolean value to set sync state to. If null, uses stored preference
     * 
     * @see startSync()
     * @see stopSync()
     */
    override suspend fun setSyncEnabled(enabled: Boolean) {
        //get default value from preferences if not set
        if (enabled  && !(ditto?.isSyncActive)!!) {
            startSync()
        } else {
            stopSync()
        }
    }

    /**
     * Starts the Ditto sync process and sets up the initial subscription.
     * 
     * This method:
     * - Initiates the sync process 
     * - Creates a subscription to the tasks collection
     * - Ensures real-time synchronization of task data across devices
     * 
     * The subscription:
     * - Queries all documents in the tasks collection
     * - Maintains the subscription until sync is stopped or the app is terminated
     * - Automatically syncs changes between the local store and the Ditto cloud
     * 
     * Implementation details:
     * - Uses Ditto's sync API to establish connection with the BigPeer
     * - Sets up a subscription using DQL to monitor the tasks collection
     * - Handles errors gracefully and logs them for debugging
     * 
     * @throws Exception if sync initialization or subscription setup fails
     *
     * Ditto Docs:
     * https://docs.ditto.live/sdk/latest/sync/start-and-stop-sync
     * https://docs.ditto.live/sdk/latest/sync/syncing-data
     *
     */
    private suspend fun startSync() {
        return withContext(Dispatchers.IO) {
            try {
                ditto?.let {
                    //
                    //TODO: implement the startSync
                    // https://docs.ditto.live/sdk/latest/install-guides/swift#integrating-and-initializing-sync
                    //

                    //UPDATE CODE HERE

                    //
                    //TODO: implement the set subscription
                    // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
                    //

                    //UPDATE CODE HERE
                }
            } catch (e: Exception) {
                errorService.showError("Failed to start ditto sync: ${e.message}")
                Log.e(TAG, "Failed to start ditto sync", e)
            }
        }
    }

    /**
     * Stops the Ditto sync process and cleans up associated resources.
     * 
     * This method:
     * - Cancels any active subscriptions to the tasks collection
     * - Stops the sync process with the Ditto cloud (BigPeer)
     * - Cleans up resources to prevent memory leaks
     * 
     * Implementation details:
     * - Closes the subscription and sets it to null
     * - Calls Ditto's stopSync() to terminate the sync process
     * - Should be called when sync is no longer needed or app is terminating
     * 
     * @see Ditto.stopSync()
     * @see DittoSyncSubscription.close()
     */
    suspend fun stopSync() {
        return withContext(Dispatchers.IO) {
            //
            //implements stopping sync by calling cancel and setting object to nil
            //and then running stopSync function
            // https://docs.ditto.live/sdk/latest/sync/syncing-data#canceling-subscriptions
            //
            subscription?.close()
            subscription = null
            ditto?.stopSync()
        }
    }



    /**
     * Creates a new TaskModel document in the Ditto store.
     * https://docs.ditto.live/sdk/latest/crud/create#creating-documents
     *
     * Implementation details:
     * - Inserts a complete document with all required fields using DQL
     *
     * @param taskModel The new TaskModel object to insert
     */
    override suspend fun insertTaskModel(taskModel: TaskModel) {
        return withContext(Dispatchers.IO) {
            try {
                //
                // TODO - write INSERT DQL Statement
                // https://docs.ditto.live/dql/insert
                // https://docs.ditto.live/sdk/latest/crud/create#creating-documents
                //

                //UPDATE CODE HERE
                val query = ""

                //
                //TODO: use dittoInstance store to execute DQL with arguments
                // https://docs.ditto.live/sdk/latest/crud/create#creating-documents
                //

                //UPDATE CODE HERE

            } catch (e: Exception) {
                errorService.showError("Failed to add taskModel: ${e.message}")
                Log.e(TAG, "Failed to add taskModel:", e)
            }
        }
    }

    /**
     * Updates existing TaskModel's properties in the Ditto store.
     * https://docs.ditto.live/sdk/latest/crud/update#updating
     * Implementation details:
     * - Uses DQL to update the document
     *
     * @param taskModel The TaskModel object containing updated values
     */
    override suspend fun updateTaskModel(taskModel: TaskModel) {
        return withContext(Dispatchers.IO) {
            try {
                //
                //TODO: write UPDATE DQL Statement
                // https://docs.ditto.live/dql/update#basic-update
                // https://docs.ditto.live/sdk/latest/crud/update#updating
                //

                //UPDATE CODE HERE
                val query = ""

                //
                //TODO: use dittoInstance store to execute DQL with arguments
                // https://docs.ditto.live/sdk/latest/crud/update#updating
                //

                //UPDATE CODE HERE

            } catch (e: Exception) {
                errorService.showError("Failed to update taskModel: ${e.message}")
                Log.e(TAG, "Failed to update taskModel:", e)
            }
        }
    }

    /**
     * Updates existing TaskModel's properties in the Ditto store.
     * https://docs.ditto.live/sdk/latest/crud/update#updating
     * Implementation details:
     * - Uses DQL to update the document
     *
     * @param taskModel The TaskModel object containing updated values
     */
    override suspend fun toggleComplete(id: String) {
        return withContext(Dispatchers.IO) {
            try {
                    //
                    //TODO: write UPDATE DQL Statement
                    // https://docs.ditto.live/dql/update#basic-update
                    // https://docs.ditto.live/sdk/latest/crud/update#updating
                    //

                    //UPDATE CODE HERE
                    val query = ""

                    //
                    //TODO: use dittoInstance store to execute DQL with arguments
                    // https://docs.ditto.live/sdk/latest/crud/update#updating
                    //

                    // UPDATE CODE HERE
                
            } catch (e: Exception) {
                errorService.showError("Failed to update taskModel: ${e.message}")
                Log.e(TAG, "Failed to update taskModel:", e)
            }
        }
    }

    /**
     * Marks a TaskModel as archived in the Ditto store.
     * This uses the 'Soft-Delete' pattern:  https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
     *
     * Implementation details:
     * - Uses a simple DQL UPDATE query to set done flag
     *
     * @param id The _id of the TaskModel to archive
     */
    override suspend fun deleteTaskModel(id: String) {
        return withContext(Dispatchers.IO) {
            try {
                //
                //TODO: write UPDATE DQL Statement using Soft-Delete pattern
                // https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
                //
                //UPDATE CODE HERE
                val query = ""


            } catch (e: Exception) {
                errorService.showError("Failed to archive taskModel: ${e.message}")
                Log.e(TAG, "Failed to archive taskModel:", e)
            }
        }
    }
}